internal/enginenetx/DESIGN.md

```diff
@@ -0,0 +1,815 @@
+# Engine Network Extensions
+
+This file documents the [./internal/enginenetx](.) package design. The content is current
+as of [probe-cli#1552](https://github.com/ooni/probe-cli/pull/1552).
+
+## Table of Contents
+
+- [Goals & Assumptions](#goals--assumptions)
+- [High-Level API](#high-level-api)
+- [Creating TLS Connections](#creating-tls-connections)
+- [Dialing Tactics](#dialing-tactics)
+- [Dialing Algorithm](#dialing-algorithm)
+- [Dialing Policies](#dialing-policies)
+	- [dnsPolicy](#dnspolicy)
+	- [userPolicy](#userpolicy)
+	- [statsPolicy](#statspolicy)
+	- [bridgePolicy](#bridgepolicy)
+- [Overall Algorithm](#overall-algorithm)
+- [Managing Stats](#managing-stats)
+- [Real-World Scenarios](#real-world-scenarios)
+	- [Invalid bridge without cached data](#invalid-bridge-without-cached-data)
+	- [Invalid bridge with cached data](#invalid-bridge-with-cached-data)
+	- [Valid bridge with invalid cached data](#valid-bridge-with-invalid-cached-data)
+- [Limitations and Future Work](#limitations-and-future-work)
+
+## Goals & Assumptions
+
+We define "bridge" an IP address with the following properties:
+
+1. the IP address is not expected to change;
+
+2. the IP address listens on port 443 and accepts _any_ incoming SNI;
+
+3. the webserver on port 443 proxies to the OONI APIs.
+
+We also assume that the Web Connectivity test helpers (TH) could accept any SNIs.
+
+We also define "tactic" a tactic to perform a TLS handshake either with a
+bridge or with a TH. We also define "policy" the collection of algorithms for
+producing tactics for performing TLS handshakes.
+
+Considering all of this, this package aims to:
+
+1. overcome DNS-based censorship for "api.ooni.io" by hardcoding known-good
+bridges IP addresses inside the codebase;
+
+2. overcome SNI-based censorship for "api.ooni.io" and test helpers by choosing
+from a pre-defined list of SNIs;
+
+3. remember and use tactics for creating TLS connections that worked previously;
+
+4. recover ~quickly if the conditions change (e.g., if a bridge is discontinued);
+
+5. adopt a censored-users-first approach where the strategy we use by default
+should allow for smooth operations _for them_ rather than prioritizing the
+non-censored case and using additional tactics as the fallback;
+
+6. try to defer sending the true `SNI` on the wire, therefore trying to
+avoid triggering potential residual censorship blocking a given TCP endpoint
+for some time regardless of what `SNI` is being used next;
+
+7. allow users to force specific bridges and SNIs by editing `$OONI_HOME/engine/bridges.conf`.
+
```

Comment:
```suggestion
8. In the event of no censorship affecting OONI API Services, the fact no tactic is working should not be a hindrance to communicating with the API. 
```

Reply:
I find this new requirement to be partially aligned with this requirement:

> 4. recover ~quickly if the conditions change (e.g., if a bridge is discontinued);

I say this because they both morally say that the system should work without big performance penalties in case there's no censorship and/or conditions change. However, your proposal for 8 is more strict and in fact looks to me like a stricter version of 4.

As such, could you propose an edit that amends 4 rather than introducing 8? Thank you!

Reply:
done

Reply:
Thanks!